From 3f00b926a11d95a6c570ff03c3a968f5f4c4cf92 Mon Sep 17 00:00:00 2001
From: Zahari Doychev <zahari.doychev@intel.com>
Date: Sat, 30 Nov 2019 10:52:30 +0100
Subject: [PATCH] dpm: support for bridge port mac learning on/off

The switchdev functionality to turn on and off the bridge port mac learning is
added to dpm.

The mac learning can be enabled/disabled by using the bridge command:

bridge link set dev pmapper4354 learning off self
bridge link set dev pmapper4354 learning on self

Signed-off-by: Zahari Doychev <zahari.doychev@intel.com>
---
 drivers/net/datapath/dpm/datapath_swdev.c          | 64 ++++++++++++++++++++--
 drivers/net/datapath/dpm/gswip31/datapath_misc.c   |  2 +
 .../net/datapath/dpm/gswip31/datapath_switchdev.c  | 58 ++++++++++++++++++++
 .../net/datapath/dpm/gswip31/datapath_switchdev.h  |  2 +
 include/net/datapath_inst.h                        |  4 ++
 5 files changed, 125 insertions(+), 5 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath_swdev.c b/drivers/net/datapath/dpm/datapath_swdev.c
index 53a62a5b7183..7598a2b0c154 100644
--- a/drivers/net/datapath/dpm/datapath_swdev.c
+++ b/drivers/net/datapath/dpm/datapath_swdev.c
@@ -730,6 +730,58 @@ int dp_del_br_if(struct net_device *dev, struct net_device *br_dev,
 	return 0;
 }
 
+static int dp_swdev_port_attr_bridge_flags_set(struct net_device *dev,
+					       unsigned long flags,
+					       struct switchdev_trans *trans)
+{
+	dp_subif_t subif = {0};
+	struct inst_info *i_info;
+	int err = 0;
+
+	if (switchdev_trans_ph_prepare(trans))
+		return 0;
+
+	if (dp_get_netif_subifid(dev, NULL, NULL, NULL, &subif, 0)) {
+		DP_DEBUG(DP_DBG_FLAG_SWDEV,
+			 "dp_get_netif_subifid failed for %s\n",
+			 dev->name);
+			return -EINVAL;
+	}
+
+	i_info = &dp_port_prop[subif.inst].info;
+	if (!i_info->swdev_bridge_port_flags_set)
+		return -EOPNOTSUPP;
+
+	err = i_info->swdev_bridge_port_flags_set(subif.inst,
+						  subif.bport,
+						  flags);
+	return err;
+}
+
+static int dp_swdev_port_attr_bridge_flags_get(struct net_device *dev,
+					       unsigned long *flags)
+{
+	dp_subif_t subif = {0};
+	struct inst_info *i_info;
+	int err = 0;
+
+	if (dp_get_netif_subifid(dev, NULL, NULL, NULL, &subif, 0)) {
+		DP_DEBUG(DP_DBG_FLAG_SWDEV,
+			 "dp_get_netif_subifid failed for %s\n",
+			 dev->name);
+			return -EINVAL;
+	}
+
+	i_info = &dp_port_prop[subif.inst].info;
+	if (!i_info->swdev_bridge_port_flags_get)
+		return -EOPNOTSUPP;
+
+	err = i_info->swdev_bridge_port_flags_get(subif.inst,
+						  subif.bport,
+						  flags);
+	return err;
+}
+
 static int dp_swdev_port_attr_set(struct net_device *dev,
 				  const struct switchdev_attr *attr,
 				  struct switchdev_trans *trans)
@@ -784,10 +836,9 @@ static int dp_swdev_port_attr_set(struct net_device *dev,
 		}
 		break;
 	case SWITCHDEV_ATTR_ID_PORT_BRIDGE_FLAGS:
-		/* err = dp_swdev_port_attr_bridge_flags_set(dp_swdev_port,
-		 *attr->u.brport_flags,
-		 *trans);
-		 */
+		err = dp_swdev_port_attr_bridge_flags_set(dev,
+							  attr->u.brport_flags,
+							  trans);
 		break;
 	case SWITCHDEV_ATTR_ID_BRIDGE_AGEING_TIME:
 		/*err = dp_swdev_port_attr_bridge_ageing_time_set(dp_swdev_port,
@@ -812,6 +863,7 @@ static int dp_swdev_port_attr_get(struct net_device *dev,
 	struct net_device *br_dev;
 	struct br_info *br_info;
 	dp_subif_t subif = {0};
+	int err = 0;
 	/*For this api default err return value "-EOPNOTSUPP"
 	 * cannot be set as this blocks bridgeport offload_fwd_mark
 	 * setting at linux bridge level("nbp_switchdev_mark_set")
@@ -843,11 +895,13 @@ static int dp_swdev_port_attr_get(struct net_device *dev,
 		//err = 0;
 		break;
 	case SWITCHDEV_ATTR_ID_PORT_BRIDGE_FLAGS:
+		err = dp_swdev_port_attr_bridge_flags_get(dev,
+							&attr->u.brport_flags);
 		break;
 	default:
 		return 0;
 	}
-	return 0;
+	return err;
 }
 
 static int dp_swdev_port_obj_add(struct net_device *dev,
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_misc.c b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
index e8f7fb6a0214..051931150082 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_misc.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_misc.c
@@ -2161,6 +2161,8 @@ int register_dp_cap_gswip31(int flag)
 	cap.info.swdev_bridge_cfg_set = dp_swdev_bridge_cfg_set;
 	cap.info.swdev_bridge_port_cfg_reset = dp_swdev_bridge_port_cfg_reset;
 	cap.info.swdev_bridge_port_cfg_set = dp_swdev_bridge_port_cfg_set;
+	cap.info.swdev_bridge_port_flags_set = dp_swdev_bridge_port_flags_set;
+	cap.info.swdev_bridge_port_flags_get = dp_swdev_bridge_port_flags_get;
 	cap.info.dp_mac_set = dp_gswip_mac_entry_add;
 	cap.info.dp_mac_reset = dp_gswip_mac_entry_del;
 	cap.info.dp_cfg_vlan = dp_gswip_ext_vlan; /*for symmetric VLAN */
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_switchdev.c b/drivers/net/datapath/dpm/gswip31/datapath_switchdev.c
index 53b90aa0b664..4423151ce871 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_switchdev.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_switchdev.c
@@ -9,6 +9,7 @@
 #include <linux/kernel.h>
 #include <linux/types.h>
 #include <linux/etherdevice.h>
+#include <linux/if_bridge.h>
 #include <net/datapath_api.h>
 #include "../datapath_swdev.h"
 #include "../datapath.h"
@@ -288,6 +289,63 @@ int dp_swdev_bridge_cfg_set(int inst, u16 fid)
 	return 0;
 }
 
+int dp_swdev_bridge_port_flags_set(int inst, int bport, unsigned long flags)
+{
+	GSW_return_t ret;
+	GSW_BRIDGE_portConfig_t bpcfg = {0};
+	struct core_ops *gsw_handle;
+	struct brdgport_ops *gsw_bp;
+	unsigned long learning;
+
+	gsw_handle = dp_port_prop[inst].ops[0];
+	gsw_bp = &gsw_handle->gsw_brdgport_ops;
+
+	bpcfg.nBridgePortId = bport;
+	bpcfg.eMask |= GSW_BRIDGE_PORT_CONFIG_MASK_MC_SRC_MAC_LEARNING;
+	ret = gsw_bp->BridgePort_ConfigGet(gsw_handle, &bpcfg);
+	if (ret != GSW_statusOk) {
+		pr_err("%s: bridge port configi get failed\r\n", __func__);
+		return -1;
+	}
+
+	learning = !bpcfg.bSrcMacLearningDisable ? BR_LEARNING : 0;
+	if ((learning ^ flags) & BR_LEARNING) {
+		bpcfg.bSrcMacLearningDisable = !bpcfg.bSrcMacLearningDisable;
+		bpcfg.eMask |= GSW_BRIDGE_PORT_CONFIG_MASK_MC_SRC_MAC_LEARNING;
+	}
+
+	ret = gsw_bp->BridgePort_ConfigSet(gsw_handle, &bpcfg);
+	if (ret != GSW_statusOk) {
+		pr_err("%s: config bridge port failed\r\n", __func__);
+		return -1;
+	}
+
+	return ret;
+}
+
+int dp_swdev_bridge_port_flags_get(int inst, int bport, unsigned long *flags)
+{
+	GSW_return_t ret;
+	GSW_BRIDGE_portConfig_t bpcfg = {0};
+	struct core_ops *gsw_handle;
+	struct brdgport_ops *gsw_bp;
+
+	gsw_handle = dp_port_prop[inst].ops[0];
+	gsw_bp = &gsw_handle->gsw_brdgport_ops;
+
+	bpcfg.nBridgePortId = bport;
+	bpcfg.eMask |= GSW_BRIDGE_PORT_CONFIG_MASK_MC_SRC_MAC_LEARNING;
+	ret = gsw_bp->BridgePort_ConfigGet(gsw_handle, &bpcfg);
+	if (ret != GSW_statusOk) {
+		pr_err("%s: bridge port configi get failed\r\n", __func__);
+		return -1;
+	}
+
+	*flags |= !bpcfg.bSrcMacLearningDisable ? BR_LEARNING : 0;
+
+	return 0;
+}
+
 int dp_swdev_free_brcfg(int inst, u16 fid)
 {
 	GSW_return_t ret;
diff --git a/drivers/net/datapath/dpm/gswip31/datapath_switchdev.h b/drivers/net/datapath/dpm/gswip31/datapath_switchdev.h
index 0f90c234d51f..8b65ce9a4115 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_switchdev.h
+++ b/drivers/net/datapath/dpm/gswip31/datapath_switchdev.h
@@ -17,4 +17,6 @@ int dp_swdev_bridge_port_cfg_reset(struct br_info *br_item,
 int dp_swdev_bridge_cfg_set(int inst, u16 fid);
 int dp_swdev_free_brcfg(int inst, u16 fid);
 int dp_gswip_ext_vlan(int inst, int vap, int ep);
+int dp_swdev_bridge_port_flags_set(int inst, int bport, unsigned long flags);
+int dp_swdev_bridge_port_flags_get(int inst, int bport, unsigned long *flags);
 #endif
diff --git a/include/net/datapath_inst.h b/include/net/datapath_inst.h
index f567a53caad2..7b27c22efa46 100644
--- a/include/net/datapath_inst.h
+++ b/include/net/datapath_inst.h
@@ -150,6 +150,10 @@ struct inst_info {
 					 int port, u32 flags);
 	int (*swdev_bridge_port_cfg_reset)(struct br_info *br_item,
 					   int inst, int bport, u16 hairpin);
+	int (*swdev_bridge_port_flags_set)(int inst, int bport,
+					  unsigned long flags);
+	int (*swdev_bridge_port_flags_get)(int inst, int bport,
+					  unsigned long *flags);
 	int (*dp_mac_set)(int bport, int fid, int inst, u8 *addr);
 	int (*dp_mac_reset)(int bport, int fid, int inst, u8 *addr);
 	int (*dp_cfg_vlan)(int inst, int vap, int ep);
